### PR TITLE
Add support for sonic and framework for search engines

### LINKED ISSUE #5096: Have a replacement for elasticsearch.
## Summary

Elasticsearch is really resource intensive and is kinda slow.
Having support for something like https://github.com/valeriansaliou/sonic which I use in my other projects and am really happy with it's performance.
It'd make it easier for people with more underpowered computers.

### PR DESCRIPTION
**Pull request by Mstrodl**
## Summary
Adds support for Sonic in addition to ElasticSearch, resolves #5096.

**Note**: I haven't been able to test ElasticSearch support as I get a ResponseError. This is probably unrelated, but it's noteworthy:
image

~~Additionally, some typings need fixing on `src/db/sonic.ts` and some small typings need to be written for 
`sonic-channel` (2 classes, 4 functions total)~~

### CODE REVIEW

**Comment by Mstrodl** [comment]
Someone smarter than me should probably make sure Elasticsearch support still works, I couldn't get a server working (inserts failed) and gave up.

**Comment by Mstrodl** [comment]
I've just written (incomplete) typings for sonic and fixed most of the typescript errors :tada:!
The remaining errors are all ones I'm not sure how to fix, but I'm sure they're trivial:
```
src/db/elasticsearch.ts(124,4): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ text: string; }'.
  No index signature with a parameter of type 'string' was found on type '{ text: string; }'.
src/db/elasticsearch.ts(124,39): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Note'.
  No index signature with a parameter of type 'string' was found on type 'Note'.
src/db/sonic.ts(109,35): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Note'.
  No index signature with a parameter of type 'string' was found on type 'Note'.
src/db/sonic.ts(152,24): error TS2339: Property 'sonic' does not exist on type 'Source & Mixin'.
src/db/sonic.ts(154,17): error TS2339: Property 'sonic' does not exist on type 'Source & Mixin'.
src/db/sonic.ts(155,17): error TS2339: Property 'sonic' does not exist on type 'Source & Mixin'.
src/db/sonic.ts(156,17): error TS2339: Property 'sonic' does not exist on type 'Source & Mixin'.
src/db/sonic.ts(156,58): error TS2339: Property 'sonic' does not exist on type 'Source & Mixin'.
```

**Comment by Mstrodl** [comment]
@acid-chicken Could you please help fix the TypeScript errors so this can be merged? I haven't ever used TS before Misskey, but most of the easy things are fixed I think

I've been using this [in production](https://social.coolmathgames.tech) for the past two weeks or so and haven't noticed any huge issues aside from some weirder stuff (not implementation issues, it's just Sonic's results are not as good as expected due to tokenization). Perhaps we should add an option in the config to optionally set the `LANG` field on Sonic requests for better results

**Comment by acid-chicken** [comment]
@Mstrodl Sure, I'll look into it.

**Comment by Mstrodl** [comment]
(Fixed merge conflict, this can be merged again)

**Comment by Mstrodl** [comment]
This should be good to merge if someone can check elasticsearch support! I fixed the typescript errors from before.

**Comment by Mstrodl** [comment]
@acid-chicken Is this something that can be merged at some point? It's been 6 months and I've been using this in my production instance with no issues.

**Review by tamaina — APPROVED** [review]

**Comment by tamaina** [comment]
https://mk.a9z.dev/ でも動かしており、しっかり日本語も検索できます。  
コードも問題なさそうですし、マージしませんか？  
@syuilo

**Comment by mei23** [comment]
> 日本語も検索できます。

前見たら日本語は使い物にならなかったです
日本語ワードは1文字単位で分割しちゃうので変なものしかかからなくて
品詞とか考慮せずにSTOPワードとして特定のワードを無条件除外しちゃうので
特定のワードを含むものが絶対にかからないはずです。
https://github.com/valeriansaliou/sonic/blob/master/src/stopwords/jpn.rs

**Comment by tamaina** [comment]
（満足に検索できるという意味ではない）

sonic側の問題なのでこのPRをマージしない理由にはならなさそう  
Elasticsearchが使えなくなるものではないわけだし

**Comment by mei23** [comment]
> （満足に検索できるという意味ではない）

しっかり日本語も検索できます って書いてあったです

> sonic側の問題なのでこのPRをマージしない理由にはならなさそう
> Elasticsearchが使えなくなるものではないわけだし

マージを否定はしてないです

**Comment by u1-liquid** [comment]
I agree with the opinion that it will not be useful

FYI: https://mk.a9z.dev/notes/8208rfgb02
https://mk.a9z.dev/search?q=%E8%A6%B3%E5%BF%B5%E7%95%8C
https://mk.a9z.dev/search?q=%E3%81%84%E3%82%89%E3%81%A3%E3%81%97%E3%82%83%E3%82%8B

**Comment by mei23** [comment]
CJK絶対に検索しないならusefulなんですけど
CJK検索する可能性があるならunusefulなんですよね

**Comment by tamaina** [comment]
https://c2.a9z.dev/ にてElasticsearchも正常に動作することを確認

（Elasticsearchもそこまで良いってわけではない気がする……設定が悪いのか？）  
https://c2.a9z.dev/search?q=%E3%83%9E%E3%83%BC%E3%82%B8

**Comment by tamaina** [comment]
検索結果にリプライが表示されてないだけだった

**Comment by Mstrodl** [comment]
@NamedKitten: Any suggestions on tuning sonic? You opened the issue saying it worked well.. Is there anything that should be done differently to make it work better?

**Review by acid-chicken — COMMENTED** [review]
@Mstrodl I'm sorry to have kept you waiting for my review.

File: src/db/SearchClientBase.ts
Comment (tamaina) [reply src/db/SearchClientBase.ts]: Why?

Comment (tamaina) [reply src/db/SearchClientBase.ts]: I'm very sorry

**Comment by tamaina** [comment]
> @NamedKitten: Any suggestions on tuning sonic? You opened the issue saying it worked well.. Is there anything that should be done differently to make it work better?

This PR works well but sonic has a problem that it is useless to search Japanese sentences.

**Comment by syuilo** [comment]
sonicのリポジトリには日本語対応してるって書いてあるけどどうなんだろ

**Comment by u1-liquid** [comment]
> sonicのリポジトリには日本語対応してるって書いてあるけどどうなんだろ

あまり検証されてないと思う

**Comment by tamaina** [comment]
sonicのコントリビューターに日本人はいなさそう

**Comment by u1-liquid** [comment]
あとそもそもSonicのレポジトリの使用例からみると全文検索に向いてるというよりはFAQとかのある程度書式がある文書のキーワード検索用なのかも

**Comment by syuilo** [comment]
まあ日本語対応してなくてもその他の言語に対応しているなら価値はあると思うのでマージ予定

**Comment by acid-chicken** [comment]
twista などで採用している MeCab search を本流にも実装して両方使うとかできてもいいかもしれない。

File: src/db/elasticsearch.ts
Comment (Mstrodl) [reply src/db/elasticsearch.ts]: Why would these ones have trailing `,`s?

**Comment by Mstrodl** [comment]
@mei23 Actually, this PR sets the locale to `none` so no lexing is done which means there should be no issue with stopwords being removed. This should probably become a configurable option at some point, but it's fine for now.

**Review by rinsuki — APPROVED** [review]

**Comment by mei23** [comment]
The `stop word` talks about sonic itself.
So it doesn't matter about this pull request.

**Comment by syuilo** [comment]
v12でマージします

**Comment by Mstrodl** [comment]
I've tested this on latest release and it seems to work fine

**Comment by u1-liquid** [comment]
hmm, since this PR does not allowing to use both ElasticSearch and Sonic, I think we don't have to separate the config section for sonic. why don't you merge the configuration with ElasticSearch?

**Comment by syuilo** [comment]
untilId方式のページングに対応するのがちょっと面倒そう

**Comment by chaoticryptidz** [comment]
Any updates when this will be merged?

**Comment by Mstrodl** [comment]
For what it's worth, I've been using this in production largely without issue

**Comment by syuilo** [comment]
TODOs for @syuilo 
- [ ] ローカルで動作確認する
- [ ] CJK対応について確認する
- [ ] IDを用いたページング対応する

**Comment by Mstrodl** [comment]
@syuilo any news? sonic is much more memory efficient than elasticsearch...

**Comment by ThatOneCalculator** [comment]
With a little bit of finagling, I was able to get this to work, even in 2022. 
(Had to `yarn add sonic-channel` )

My server's RAM went down from 70% usage to 14% usage. This is amazing!

@syuilo

**Comment by syuilo** [comment]
It only supports some languages, and there is a problem that paging using IDs is not possible, but the latter is a similar problem with the current ElasticSearch, so it seems like it should be fixed in the future.
I' ll merge this after resolving the conflicts.

**Comment by mei23** [comment]
外部エンジンは、DB自体の検索と違って ミュート/ブロック/削除 に追従できない問題があるけど、それはどうする？

**Comment by syuilo** [comment]
Sonicから返ってくるのはIDだけみたいだから、最終的にはPostgreSQLにクエリすることになる
その際にミュートなど考慮するようにすれば良いと思った

**Comment by mei23** [comment]
つまり (Deleted) (Muted) みたいな、エントリを表示したりしなかったりになると…

**Comment by syuilo** [comment]
> エントリを表示したりしなかったり

がよくわからなかった

**Comment by Mstrodl** [comment]
> It only supports some languages, and there is a problem that paging using IDs is not possible, but the latter is a similar problem with the current ElasticSearch, so it seems like it should be fixed in the future. I' ll merge this after resolving the conflicts.

I probably won't fix the conflicts, I haven't been using Misskey in the past few months because it became difficult to keep my fork updated (to keep my sonic changes) and I don't use the fediverse often

If someone is interested in cleaning up the conflicts they are certainly free to do so, but I don't have the time or motivation to do it myself

**Comment by ThatOneCalculator** [comment]
> If someone is interested in cleaning up the conflicts they are certainly free to do so, but I don't have the time or motivation to do it myself

I got it working on my instance! Give me about a day and I can open up a new PR to get it working :D

**Comment by ThatOneCalculator** [comment]
@Mstrodl @syuilo see #8299 :)

**Comment by tamaina** [comment]
Sorry, this PR is closed due to incompatibility with the directory structure.

**Pull request closed without merging**
